File: frontend/src/schemas/knowledgeBase.ts
import { z } from 'zod';

import { explanationLevel } from '@/schemas/enums';
import { calendarDate, instant, uuid } from '@/schemas/primitives';

/**
 * Knowledge Base requests and responses.
 *
 * Every optional field is `.nullable()` and not `.optional()`, which is the shape the
 * backend actually sends: it serialises an unset field as an explicit null rather than
 * leaving the key out. Marking them optional would type them as possibly-absent and,
 * under this project's strict optional-property setting, force callers to handle an
 * `undefined` that never arrives while still leaving the null they will actually get.
 */

export const knowledgeBase = z.object({
  id: uuid,
  user_id: uuid,
  name: z.string(),
  description: z.string().nullable(),
  subject: z.string().nullable(),
  learning_goal: z.string().nullable(),
  preferred_language: z.string(),
  explanation_level: explanationLevel,
  exam_date: calendarDate.nullable(),
  graph_enabled: z.boolean(),
  active_index_version: z.number().int(),
  active_graph_version: z.number().int(),
  created_at: instant,
  updated_at: instant,
});
export type KnowledgeBase = z.infer<typeof knowledgeBase>;

export const knowledgeBaseList = z.array(knowledgeBase);

/**
 * What creating one accepts. The name bound is the backend's, repeated here so a name
 * that is going to be refused is refused before a request is spent finding that out.
 */
export const createKnowledgeBaseRequest = z.object({
  name: z.string().min(1).max(200),
  description: z.string().nullable().optional(),
  subject: z.string().nullable().optional(),
  learning_goal: z.string().nullable().optional(),
  preferred_language: z.string().nullable().optional(),
  explanation_level: explanationLevel.nullable().optional(),
  exam_date: calendarDate.nullable().optional(),
});
export type CreateKnowledgeBaseRequest = z.input<typeof createKnowledgeBaseRequest>;

/** Every field is optional on update, including the name, which may not be blanked. */
export const updateKnowledgeBaseRequest = createKnowledgeBaseRequest.partial();
export type UpdateKnowledgeBaseRequest = z.input<typeof updateKnowledgeBaseRequest>;

/**
 * What a queued rebuild reports.
 *
 * `active_index_version` is the version retrieval still answers from, and it stays that
 * way until every document has been read again — so a caller watching this number knows
 * the rebuild landed when it changes, rather than when the job was accepted.
 */
export const reindexResponse = z.object({
  knowledge_base_id: uuid,
  job_id: uuid,
  documents: z.number().int(),
  active_index_version: z.number().int(),
  target_index_version: z.number().int(),
});
export type ReindexResponse = z.infer<typeof reindexResponse>;

File: backend/app/api/schemas/knowledge_base.py
"""Pydantic schemas for the Knowledge Base CRUD API."""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.domain.enums import ExplanationLevel


class CreateKnowledgeBaseRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    subject: str | None = None
    learning_goal: str | None = None
    preferred_language: str | None = None
    explanation_level: ExplanationLevel | None = None
    exam_date: date | None = None


class UpdateKnowledgeBaseRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    subject: str | None = None
    learning_goal: str | None = None
    preferred_language: str | None = None
    explanation_level: ExplanationLevel | None = None
    exam_date: date | None = None
    graph_enabled: bool | None = None


class KnowledgeBaseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    name: str
    description: str | None = None
    subject: str | None = None
    learning_goal: str | None = None
    preferred_language: str
    explanation_level: ExplanationLevel
    exam_date: date | None = None
    graph_enabled: bool
    active_index_version: int
    active_graph_version: int
    created_at: datetime
    updated_at: datetime


class ReindexResponse(BaseModel):
    """What was queued, and what retrieval is answering from until it finishes."""

    knowledge_base_id: UUID
    job_id: UUID
    documents: int
    #: The version retrieval still answers from. It stays this until every document has
    #: been read again, so a caller polling the Knowledge Base can see when the rebuild
    #: landed by watching this number change.
    active_index_version: int
    target_index_version: int


class BuildGraphResponse(BaseModel):
    """How many graph-build jobs were queued."""

    knowledge_base_id: UUID
    jobs_queued: int
